File: system-test/dotnet/Dsl.Core/Scenario/Assume/AssumeStage.cs
using Dsl.Port.Assume;
using Dsl.Port.Assume.Steps;
using Optivem.Testing;

namespace Dsl.Core.Scenario.Assume;

public class AssumeStage : IAssumeStage
{
    private readonly UseCaseDsl _app;
    private readonly Channel? _channel;

    public AssumeStage(UseCaseDsl app, Channel? channel = null)
    {
        _app = app;
        _channel = channel;
    }

    public IAssumeRunning MyShop()
    {
        return new AssumeRunningAction(async () =>
        {
            (await (await _app.MyShop(_channel)).GoToMyShop().Execute()).ShouldSucceed();
        }, this);
    }

    public IAssumeRunning Erp()
    {
        return new AssumeRunningAction(async () =>
        {
            (await _app.Erp().GoToErp().Execute()).ShouldSucceed();
        }, this);
    }

    public IAssumeRunning Tax()
    {
        return new AssumeRunningAction(async () =>
        {
            (await _app.Tax().GoToTax().Execute()).ShouldSucceed();
        }, this);
    }

    public IAssumeRunning Clock()
    {
        return new AssumeRunningAction(async () =>
        {
            (await _app.Clock().GoToClock().Execute()).ShouldSucceed();
        }, this);
    }

    private class AssumeRunningAction : IAssumeRunning
    {
        private readonly Func<Task> _action;
        private readonly IAssumeStage _assumeStage;

        public AssumeRunningAction(Func<Task> action, IAssumeStage assumeStage)
        {
            _action = action;
            _assumeStage = assumeStage;
        }

        public async Task<IAssumeStage> ShouldBeRunning()
        {
            await _action();
            return _assumeStage;
        }
    }
}

File: system-test/typescript/src/testkit/dsl/core/scenario/assume/assume-stage.ts
import { expect } from '@playwright/test';
import { AppContext } from '../app-context.js';
import type { AssumeStage as IAssumeStage } from '../../../port/assume/assume-stage.js';
import type { AssumeRunning as IAssumeRunning } from '../../../port/assume/steps/assume-running.js';

export class AssumeStage implements IAssumeStage {
  constructor(private readonly app: AppContext) {}

  myShop(): AssumeRunning {
    return new AssumeRunning(this, async () => {
      const result = await this.app.myShop().goToMyShop();
      expect(result.success).toBe(true);
    });
  }

  erp(): AssumeRunning {
    return new AssumeRunning(this, async () => {
      const result = await this.app.erpDriver.goToErp();
      expect(result.success).toBe(true);
    });
  }

  clock(): AssumeRunning {
    return new AssumeRunning(this, async () => {
      const result = await this.app.clockDriver.goToClock();
      expect(result.success).toBe(true);
    });
  }

  tax(): AssumeRunning {
    return new AssumeRunning(this, async () => {
      const result = await this.app.taxDriver.goToTax();
      expect(result.success).toBe(true);
    });
  }
}

export class AssumeRunning implements IAssumeRunning {
  constructor(
    private readonly stage: AssumeStage,
    private readonly checkFn: () => Promise<void>,
  ) {}

  async shouldBeRunning(): Promise<AssumeStage> {
    await this.checkFn();
    return this.stage;
  }
}
